### PR TITLE
[14][BACKPORT] stock_available_to_promise add unrelease

### PR DESCRIPTION
**Pull request by TDu**
This is from a pr on 16.0

* https://github.com/OCA/wms/pull/478

A unit test with a combo of 2 steps delivery is being added.
And also unreleasing a chain of moves on a move cancellation.

For this to work the following backport fix on Odoo is needed:
* https://github.com/odoo/odoo/pull/111668
It is merged now :heavy_check_mark: 

And this PR to avoid conflict with the previous mentioned pr
* https://github.com/OCA/stock-logistics-workflow/pull/1222

ref.: rau-114

### CODE REVIEW

File: stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def migrate(cr, version):
+    _logger.info(
+        "stock_available_to_promise_release: Configure picking types"
+        " to prevent new moves into released pickings"
+    )
+    cr.execute("update stock_picking_type set prevent_new_move_after_release = true")
+    _logger.info(
+        "stock_available_to_promise_release: Set last release date on" " some pickings"
```
Comment (jbaudoux) [inline stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py]:
```suggestion
        "stock_available_to_promise_release: Set last release date on some pickings"
```

File: stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        left join stock_rule sr on sr.id = sm.rule_id
+        where sp.printed = true and spt.code = 'outgoing'
+            and sr.available_to_promise_defer_pull = true
+        group by sp.id
+)
+update stock_picking
+    set last_release_date = create_date
+    from picks
+    where stock_picking.id = picks.id
+    """
```
Comment (jbaudoux) [inline stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py:34]: maybe safer to add
```suggestion
    and not last_release_date"""
```

File: stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py (hunk 1 above)
Comment (TDu) [reply stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py]: fixed

File: stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py (hunk 2 above)
Comment (TDu) [reply stock_available_to_promise_release/migrations/14.0.2.0.0/post-migrate.py:34]: done

**Review by jbaudoux — APPROVED** [review]
Tested and working

**Comment by jbaudoux** [comment]
@TDu Can you squash last commits so that we can merge it?

File: stock_available_to_promise_release/tests/__init__.py [hunk 3]
```diff
@@ -1 +1,3 @@
 from . import test_reservation
+from . import test_unrelease
+from . import test_unrelease_2steps
```
Comment (jbaudoux) [inline stock_available_to_promise_release/tests/__init__.py:3]: Your new test is not included
```suggestion
from . import test_unrelease_2steps
from . import test_unrelease_3steps
```

Comment (TDu) [reply stock_available_to_promise_release/tests/__init__.py:3]: done

**Review by lmignon — APPROVED** [review]
LGTM (Code review)

File: stock_available_to_promise_release/models/stock_picking_type.py [hunk 4]
```diff
@@ -8,6 +8,17 @@ class StockPickingType(models.Model):
     _inherit = "stock.picking.type"
 
     count_picking_need_release = fields.Integer(compute="_compute_picking_count")
+    unrelease_on_backorder = fields.Boolean(
+        string="Unrelease on backorder",
+        help="If checked, when a backorder is created the moves into the "
+        "backorder are unreleased if they come from a route configured "
+        "to manually release moves",
+    )
+    prevent_new_move_after_release = fields.Boolean(
```
Comment (sebalix) [inline stock_available_to_promise_release/models/stock_picking_type.py:17]: Should we add `default=True` on this parameter?

**Review by sebalix — APPROVED** [review]

**Comment by TDu** [comment]
Just a rebase...

**Pull request merged**